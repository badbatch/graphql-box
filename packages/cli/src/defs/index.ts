import { PlainObjectMap } from "@graphql-box/core";

export interface IntrospectArgs {
  headers?: string[];
  output: string;
  schemaPath?: string;
  url?: string;
}

export interface IntrospectionResult {
  data?: PlainObjectMap;
  [key: string]: any;
}
