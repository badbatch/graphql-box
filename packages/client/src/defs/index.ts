import { CacheManagerDef, CacheManagerInit } from "@graphql-box/cache-manager";
import {
  DebugManagerDef,
  MaybeRequestResult,
  RequestContext,
  RequestManagerDef,
  RequestManagerInit,
  RequestOptions,
  SubscriptionsManagerDef,
  SubscriptionsManagerInit,
} from "@graphql-box/core";
import { DebugManagerInit } from "@graphql-box/debug-manager";
import { RequestParserDef, RequestParserInit } from "@graphql-box/request-parser";

export interface UserOptions {
  /**
   * The curried function to initialize the cache manager.
   */
  cacheManager: CacheManagerInit;

  /**
   * The curried function to initialize the request manager.
   */
  requestManager: RequestManagerInit;

  /**
   * The curried function to initialize the debug manager.
   */
  debugManager?: DebugManagerInit;

  /**
   * The curried function to initialzie the request parser.
   */
  requestParser: RequestParserInit;

  /**
   * The curried function to initialize the subscriptions manager.
   */
  subscriptionsManager?: SubscriptionsManagerInit;

  /**
   * The name of the property thats value is used as the unique
   * identifier for each type in the GraphQL schema.
   */
  typeIDKey?: string;
}

export interface ConstructorOptions {
  /**
   * The cache manager.
   */
  cacheManager: CacheManagerDef;

  /**
   * The debug manager.
   */
  debugManager?: DebugManagerDef;

  /**
   * The request manager.
   */
  requestManager: RequestManagerDef;

  /**
   * The GraphQL request parser.
   */
  requestParser: RequestParserDef;

  /**
   * The subscriptions manager.
   */
  subscriptionsManager?: SubscriptionsManagerDef;
}

export type PendingQueryResolver = (value: MaybeRequestResult) => void;

export interface PendingQueryData {
  context: RequestContext;
  options: RequestOptions;
  resolve: PendingQueryResolver;
}

export interface QueryTracker {
  active: string[];
  pending: Map<string, PendingQueryData[]>;
}
