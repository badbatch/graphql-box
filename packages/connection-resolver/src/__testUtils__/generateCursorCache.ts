import Cachemap from "@cachemap/core";
import map from "@cachemap/map";
import { encode } from "js-base64";
import cacheCursors from "../helpers/cacheCursors";
import generatePages from "./generatePages";

export type Params = {
  group: string;
  pageRanges?: string[];
  resultsPerPage?: number;
  totalPages: number;
  totalResults: number;
};

export default async ({ group, pageRanges = [], resultsPerPage, totalPages, totalResults }: Params) => {
  const cursorCache = new Cachemap({
    name: "cursorCache",
    store: map(),
  });

  const headers = new Headers({ "Cache-Control": "max-age=60" });

  if (pageRanges.length && resultsPerPage) {
    const pages = generatePages(pageRanges);

    await Promise.all(
      pages.map(async page => {
        const isLastPage = page === pages[pages.length - 1];
        let resultsOnCurrentPage: number;

        if (isLastPage) {
          const remainder = totalResults % resultsPerPage;
          resultsOnCurrentPage = remainder === 0 ? resultsPerPage : remainder;
        } else {
          resultsOnCurrentPage = resultsPerPage;
        }

        const edges = Array.from({ length: resultsOnCurrentPage }, (_v, i) => i).map(index => {
          const id = encode(`${index}::${page}`);
          return { cursor: `${id}::${group}`, node: { id } };
        });

        await cacheCursors(cursorCache, { edges, group, headers, page, totalPages, totalResults });
      }),
    );
  } else {
    await cursorCache.set(
      `${group}-metadata`,
      { totalPages, totalResults },
      { cacheHeaders: { cacheControl: headers.get("cache-control") ?? undefined } },
    );
  }

  return cursorCache;
};
