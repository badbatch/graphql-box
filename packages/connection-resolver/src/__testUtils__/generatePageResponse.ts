import { type PlainObject } from '@graphql-box/core';
import { encode } from 'js-base64';
import { type ResourceResponse } from '../types.ts';

export type Params = {
  resultsPerPage: number;
  totalPages: number;
  totalResults: number;
};

export const generatePageResponse =
  ({ resultsPerPage, totalPages, totalResults }: Params) =>
  (page: number) =>
    Promise.resolve({
      data: {
        page,
        results: Array.from({ length: resultsPerPage }, (_v, index) => index).map(index => {
          return { id: encode(`${index}::${page}`) };
        }),
        totalPages,
        totalResults,
      },
      headers: new Headers({ 'Cache-Control': 'max-age=60' }),
    }) as unknown as Promise<ResourceResponse<PlainObject>>;
