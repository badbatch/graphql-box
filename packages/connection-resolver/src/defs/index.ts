import Cachemap from "@cachemap/core";
import { GraphQLResolveInfo } from "graphql";

export type PlainObject = {
  [key: string]: any;
};

export type Node = PlainObject & { id: string };

export type Direction = "backward" | "forward";

export type CursorCacheEntry = {
  group: string;
  index: number;
  node: Node;
  page: number;
};

export type CursorGroupMetadata = {
  totalPages: number;
  totalResults: number;
};

export type CachedEdges = {
  edges: Edge[];
  pageNumber: number;
};

export type Edge = {
  cursor: string;
  node: Node;
};

export type Indexes = {
  absolute: number;
  relative: number;
};

export type PageInfo = {
  endCursor?: string;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  startCursor?: string;
};

export type Connection = {
  edges: Edge[];
  errors: Error[];
  nodes: Node[];
  pageInfo: PageInfo;
  totalCount: number;
};

export interface ResourceResponse extends Response {
  data?: PlainObject;
  errors?: Error[];
}

export type ResourceResolver = (args: { page: number }) => Promise<ResourceResponse>;

export type CreateResourceResolver<Source extends PlainObject, Args extends PlainObject, Ctx extends PlainObject> = (
  source: Source,
  args: Args,
  context: Ctx,
  info: GraphQLResolveInfo,
) => ResourceResolver;

export interface Getters<Resource extends PlainObject, ResourceNode extends Node> {
  nodes: (obj: Resource) => ResourceNode[];
  page: (obj: Resource) => number;
  totalPages: (obj: Resource) => number;
  totalResults: (obj: Resource) => number;
}

export interface ConnectionResolverUserOptions<
  Source extends PlainObject,
  Args extends PlainObject,
  Ctx extends PlainObject,
  Resource extends PlainObject,
  ResourceNode extends Node
> {
  createMakeCursors: (
    source: Source,
    args: Args,
    context: Ctx,
    info: GraphQLResolveInfo,
  ) => {
    makeGroupCursor: () => string;
    makeIDCursor: (id: string | number) => string;
  };
  createResourceResolver: CreateResourceResolver<Source, Args, Ctx>;
  cursorCache: Cachemap;
  getters: Getters<Resource, ResourceNode>;
  resolver?: (args: Connection) => Connection;
  resultsPerPage: number;
}

export type ConnectionResolver = (
  source: PlainObject,
  args: PlainObject & ConnectionInputOptions,
  context: PlainObject,
  info: GraphQLResolveInfo,
) => Promise<Connection>;

export type ConnectionInputOptions = {
  after?: string;
  before?: string;
  first?: number;
  last?: number;
};

export type Context = {
  entry: CursorCacheEntry;
  metadata: CursorGroupMetadata;
  resultsPerPage: number;
};
