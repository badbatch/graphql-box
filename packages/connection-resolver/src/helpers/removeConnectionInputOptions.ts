import { pickBy } from "lodash";

const connectionInputOptions = ["after", "before", "first", "last"];

export default <O extends object>(args: O) => pickBy(args, (_value, key) => !connectionInputOptions.includes(key));
