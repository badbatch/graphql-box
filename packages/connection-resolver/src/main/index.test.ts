import Cachemap from "@cachemap/core";
import map from "@cachemap/map";
import { GraphQLError, GraphQLResolveInfo } from "graphql";
import { encode } from "js-base64";
import makeConnectionResolver from ".";
import generateCursorCache from "../__testUtils__/generateCursorCache";
import generatePageResponse from "../__testUtils__/generatePageResponse";
import { PlainObject } from "../defs";
import removeConnectionInputOptions from "../helpers/removeConnectionInputOptions";

describe("connectionResolver", () => {
  const createMakeCursors = (_source: PlainObject, args: PlainObject) => ({
    makeGroupCursor: () => encode(JSON.stringify(removeConnectionInputOptions(args))),
    makeIDCursor: (id: string | number) => encode(`${id}::${JSON.stringify(removeConnectionInputOptions(args))}`),
  });

  const getters = {
    nodes: ({ results }: PlainObject) => results,
    page: ({ page }: PlainObject) => page,
    totalPages: ({ totalPages }: PlainObject) => totalPages,
    totalResults: ({ totalResults }: PlainObject) => totalResults,
  };

  const resultsPerPage = 10;

  describe("when a cursor is supplied", () => {
    test("when the cursor is invalid", async () => {
      const createResourceResolver = (_obj: PlainObject, args: PlainObject, { restClient }: PlainObject) => {
        return async ({ page }: { page: number }) => restClient({ ...removeConnectionInputOptions(args), page });
      };

      const connectionResolver = makeConnectionResolver({
        createMakeCursors,
        createResourceResolver,
        cursorCache: new Cachemap({
          name: "GRAPHQL_BOX_CONNECTION_RESOLVER",
          store: map(),
        }),
        getters,
        resultsPerPage,
      });

      const args = { before: "abcdefg", first: 5, query: "Hello world!" };
      const info = {};

      try {
        await connectionResolver({}, args, {}, info as GraphQLResolveInfo);
      } catch (e) {
        expect(e).toBeInstanceOf(GraphQLError);
        expect(e.message.startsWith("Invalid connection argument combination")).toBe(true);
      }
    });

    test("when there are NO missing pages in the cache", async () => {
      const createResourceResolver = (_obj: PlainObject, args: PlainObject, { restClient }: PlainObject) => {
        return async ({ page }: { page: number }) => restClient({ ...removeConnectionInputOptions(args), page });
      };

      const groupCursor = encode(JSON.stringify({ query: "Hello world!" }));

      const cursorCache = await generateCursorCache({
        group: groupCursor,
        pageRanges: ["1-10"],
        resultsPerPage,
        totalPages: 10,
        totalResults: 100,
      });

      const connectionResolver = makeConnectionResolver({
        createMakeCursors,
        createResourceResolver,
        cursorCache,
        getters,
        resultsPerPage,
      });

      const args = {
        after: `${encode(`0::1`)}::${groupCursor}`,
        first: 5,
        query: "Hello world!",
      };

      const info = {};
      expect(await connectionResolver({}, args, {}, info as GraphQLResolveInfo)).toMatchSnapshot();
    });

    test("when there are missing pages in the cache", async () => {
      const pageResponse = generatePageResponse({ resultsPerPage, totalPages: 10, totalResults: 100 });
      const mock = jest.fn().mockImplementation(page => pageResponse(page));

      const createResourceResolver = (
        _obj: PlainObject,
        _args: PlainObject,
        _context: PlainObject,
        _info: GraphQLResolveInfo,
      ) => async ({ page }: { page: number }) => mock(page);

      const groupCursor = encode(JSON.stringify({ query: "Hello world!" }));

      const cursorCache = await generateCursorCache({
        group: groupCursor,
        pageRanges: ["1", "3-10"],
        resultsPerPage,
        totalPages: 10,
        totalResults: 100,
      });

      const connectionResolver = makeConnectionResolver({
        createMakeCursors,
        createResourceResolver,
        cursorCache,
        getters,
        resultsPerPage,
      });

      const args = {
        after: `${encode(`0::1`)}::${groupCursor}`,
        first: 35,
        query: "Hello world!",
      };

      const info = {};
      expect(await connectionResolver({}, args, {}, info as GraphQLResolveInfo)).toMatchSnapshot();
      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenCalledWith(2);
    });
  });

  describe("when the first [X] number are requested", () => {
    test("when there is a fresh cache and there are NO missing pages in the cache", async () => {
      const createResourceResolver = (_obj: PlainObject, args: PlainObject, { restClient }: PlainObject) => {
        return async ({ page }: { page: number }) => restClient({ ...removeConnectionInputOptions(args), page });
      };

      const groupCursor = encode(JSON.stringify({ query: "Hello world!" }));

      const cursorCache = await generateCursorCache({
        group: groupCursor,
        pageRanges: ["1-10"],
        resultsPerPage,
        totalPages: 10,
        totalResults: 100,
      });

      const connectionResolver = makeConnectionResolver({
        createMakeCursors,
        createResourceResolver,
        cursorCache,
        getters,
        resultsPerPage,
      });

      const args = {
        first: 5,
        query: "Hello world!",
      };

      const info = {};
      expect(await connectionResolver({}, args, {}, info as GraphQLResolveInfo)).toMatchSnapshot();
    });

    test("when there is a fresh cache and there are missing pages in the cache", async () => {
      const pageResponse = generatePageResponse({ resultsPerPage, totalPages: 10, totalResults: 100 });
      const mock = jest.fn().mockImplementation(page => pageResponse(page));

      const createResourceResolver = (
        _obj: PlainObject,
        _args: PlainObject,
        _context: PlainObject,
        _info: GraphQLResolveInfo,
      ) => async ({ page }: { page: number }) => mock(page);

      const groupCursor = encode(JSON.stringify({ query: "Hello world!" }));

      const cursorCache = await generateCursorCache({
        group: groupCursor,
        pageRanges: ["2-10"],
        resultsPerPage,
        totalPages: 10,
        totalResults: 100,
      });

      const connectionResolver = makeConnectionResolver({
        createMakeCursors,
        createResourceResolver,
        cursorCache,
        getters,
        resultsPerPage,
      });

      const args = {
        first: 5,
        query: "Hello world!",
      };

      const info = {};
      expect(await connectionResolver({}, args, {}, info as GraphQLResolveInfo)).toMatchSnapshot();
      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenCalledWith(1);
    });

    test("when there is NOT a fresh cache", async () => {
      const pageResponse = generatePageResponse({ resultsPerPage, totalPages: 10, totalResults: 100 });
      const mock = jest.fn().mockImplementation(page => pageResponse(page));

      const createResourceResolver = (
        _obj: PlainObject,
        _args: PlainObject,
        _context: PlainObject,
        _info: GraphQLResolveInfo,
      ) => async ({ page }: { page: number }) => mock(page);

      const cursorCache = new Cachemap({
        name: "cursorCache",
        store: map(),
      });

      const connectionResolver = makeConnectionResolver({
        createMakeCursors,
        createResourceResolver,
        cursorCache,
        getters,
        resultsPerPage,
      });

      const args = {
        first: 5,
        query: "Hello world!",
      };

      const info = {};
      expect(await connectionResolver({}, args, {}, info as GraphQLResolveInfo)).toMatchSnapshot();
      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenCalledWith(1);
    });
  });

  describe("when the last [X] number are requested", () => {
    test("when there is a fresh cache and there are NO missing pages in the cache", async () => {
      const createResourceResolver = (_obj: PlainObject, args: PlainObject, { restClient }: PlainObject) => {
        return async ({ page }: { page: number }) => restClient({ ...removeConnectionInputOptions(args), page });
      };

      const groupCursor = encode(JSON.stringify({ query: "Hello world!" }));

      const cursorCache = await generateCursorCache({
        group: groupCursor,
        pageRanges: ["1-10"],
        resultsPerPage,
        totalPages: 10,
        totalResults: 100,
      });

      const connectionResolver = makeConnectionResolver({
        createMakeCursors,
        createResourceResolver,
        cursorCache,
        getters,
        resultsPerPage,
      });

      const args = {
        last: 5,
        query: "Hello world!",
      };

      const info = {};
      expect(await connectionResolver({}, args, {}, info as GraphQLResolveInfo)).toMatchSnapshot();
    });

    test("when there is a fresh cache and there are missing pages in the cache", async () => {
      const pageResponse = generatePageResponse({ resultsPerPage, totalPages: 10, totalResults: 100 });
      const mock = jest.fn().mockImplementation(page => pageResponse(page));

      const createResourceResolver = (
        _obj: PlainObject,
        _args: PlainObject,
        _context: PlainObject,
        _info: GraphQLResolveInfo,
      ) => async ({ page }: { page: number }) => mock(page);

      const groupCursor = encode(JSON.stringify({ query: "Hello world!" }));

      const cursorCache = await generateCursorCache({
        group: groupCursor,
        pageRanges: ["1-9"],
        resultsPerPage,
        totalPages: 10,
        totalResults: 100,
      });

      const connectionResolver = makeConnectionResolver({
        createMakeCursors,
        createResourceResolver,
        cursorCache,
        getters,
        resultsPerPage,
      });

      const args = {
        last: 5,
        query: "Hello world!",
      };

      const info = {};
      expect(await connectionResolver({}, args, {}, info as GraphQLResolveInfo)).toMatchSnapshot();
      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenCalledWith(10);
    });

    test("when there is NOT a fresh cache", async () => {
      const pageResponse = generatePageResponse({ resultsPerPage, totalPages: 10, totalResults: 100 });
      const mock = jest.fn().mockImplementation(page => pageResponse(page));

      const createResourceResolver = (
        _obj: PlainObject,
        _args: PlainObject,
        _context: PlainObject,
        _info: GraphQLResolveInfo,
      ) => async ({ page }: { page: number }) => mock(page);

      const cursorCache = new Cachemap({
        name: "cursorCache",
        store: map(),
      });

      const connectionResolver = makeConnectionResolver({
        createMakeCursors,
        createResourceResolver,
        cursorCache,
        getters,
        resultsPerPage,
      });

      const args = {
        last: 5,
        query: "Hello world!",
      };

      const info = {};
      expect(await connectionResolver({}, args, {}, info as GraphQLResolveInfo)).toMatchSnapshot();
      expect(mock).toHaveBeenCalledTimes(2);
      expect(mock).toHaveBeenCalledWith(1);
      expect(mock).toHaveBeenCalledWith(10);
    });
  });
});
