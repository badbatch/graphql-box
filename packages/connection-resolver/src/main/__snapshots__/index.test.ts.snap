// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectionResolver when a cursor is supplied when there are NO missing pages in the cache 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "MTo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "MTo6MQ==",
      },
    },
    Object {
      "cursor": "Mjo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Mjo6MQ==",
      },
    },
    Object {
      "cursor": "Mzo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Mzo6MQ==",
      },
    },
    Object {
      "cursor": "NDo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "NDo6MQ==",
      },
    },
    Object {
      "cursor": "NTo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "NTo6MQ==",
      },
    },
  ],
  "errors": Array [],
  "nodes": Array [
    Object {
      "id": "MTo6MQ==",
    },
    Object {
      "id": "Mjo6MQ==",
    },
    Object {
      "id": "Mzo6MQ==",
    },
    Object {
      "id": "NDo6MQ==",
    },
    Object {
      "id": "NTo6MQ==",
    },
  ],
  "pageInfo": Object {
    "endCursor": "NTo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
    "hasNextPage": true,
    "hasPreviousPage": true,
    "startCursor": "MTo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
  },
  "totalCount": 100,
}
`;

exports[`connectionResolver when a cursor is supplied when there are missing pages in the cache 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "MTo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "MTo6MQ==",
      },
    },
    Object {
      "cursor": "Mjo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Mjo6MQ==",
      },
    },
    Object {
      "cursor": "Mzo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Mzo6MQ==",
      },
    },
    Object {
      "cursor": "NDo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "NDo6MQ==",
      },
    },
    Object {
      "cursor": "NTo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "NTo6MQ==",
      },
    },
    Object {
      "cursor": "Njo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Njo6MQ==",
      },
    },
    Object {
      "cursor": "Nzo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Nzo6MQ==",
      },
    },
    Object {
      "cursor": "ODo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "ODo6MQ==",
      },
    },
    Object {
      "cursor": "OTo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "OTo6MQ==",
      },
    },
    Object {
      "cursor": "TURvNk1nPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "MDo6Mg==",
      },
    },
    Object {
      "cursor": "TVRvNk1nPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "MTo6Mg==",
      },
    },
    Object {
      "cursor": "TWpvNk1nPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "Mjo6Mg==",
      },
    },
    Object {
      "cursor": "TXpvNk1nPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "Mzo6Mg==",
      },
    },
    Object {
      "cursor": "TkRvNk1nPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "NDo6Mg==",
      },
    },
    Object {
      "cursor": "TlRvNk1nPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "NTo6Mg==",
      },
    },
    Object {
      "cursor": "TmpvNk1nPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "Njo6Mg==",
      },
    },
    Object {
      "cursor": "TnpvNk1nPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "Nzo6Mg==",
      },
    },
    Object {
      "cursor": "T0RvNk1nPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "ODo6Mg==",
      },
    },
    Object {
      "cursor": "T1RvNk1nPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "OTo6Mg==",
      },
    },
    Object {
      "cursor": "MDo6Mw==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "MDo6Mw==",
      },
    },
    Object {
      "cursor": "MTo6Mw==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "MTo6Mw==",
      },
    },
    Object {
      "cursor": "Mjo6Mw==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Mjo6Mw==",
      },
    },
    Object {
      "cursor": "Mzo6Mw==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Mzo6Mw==",
      },
    },
    Object {
      "cursor": "NDo6Mw==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "NDo6Mw==",
      },
    },
    Object {
      "cursor": "NTo6Mw==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "NTo6Mw==",
      },
    },
    Object {
      "cursor": "Njo6Mw==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Njo6Mw==",
      },
    },
    Object {
      "cursor": "Nzo6Mw==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Nzo6Mw==",
      },
    },
    Object {
      "cursor": "ODo6Mw==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "ODo6Mw==",
      },
    },
    Object {
      "cursor": "OTo6Mw==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "OTo6Mw==",
      },
    },
    Object {
      "cursor": "MDo6NA==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "MDo6NA==",
      },
    },
    Object {
      "cursor": "MTo6NA==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "MTo6NA==",
      },
    },
    Object {
      "cursor": "Mjo6NA==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Mjo6NA==",
      },
    },
    Object {
      "cursor": "Mzo6NA==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Mzo6NA==",
      },
    },
    Object {
      "cursor": "NDo6NA==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "NDo6NA==",
      },
    },
    Object {
      "cursor": "NTo6NA==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "NTo6NA==",
      },
    },
  ],
  "errors": Array [],
  "nodes": Array [
    Object {
      "id": "MTo6MQ==",
    },
    Object {
      "id": "Mjo6MQ==",
    },
    Object {
      "id": "Mzo6MQ==",
    },
    Object {
      "id": "NDo6MQ==",
    },
    Object {
      "id": "NTo6MQ==",
    },
    Object {
      "id": "Njo6MQ==",
    },
    Object {
      "id": "Nzo6MQ==",
    },
    Object {
      "id": "ODo6MQ==",
    },
    Object {
      "id": "OTo6MQ==",
    },
    Object {
      "id": "MDo6Mg==",
    },
    Object {
      "id": "MTo6Mg==",
    },
    Object {
      "id": "Mjo6Mg==",
    },
    Object {
      "id": "Mzo6Mg==",
    },
    Object {
      "id": "NDo6Mg==",
    },
    Object {
      "id": "NTo6Mg==",
    },
    Object {
      "id": "Njo6Mg==",
    },
    Object {
      "id": "Nzo6Mg==",
    },
    Object {
      "id": "ODo6Mg==",
    },
    Object {
      "id": "OTo6Mg==",
    },
    Object {
      "id": "MDo6Mw==",
    },
    Object {
      "id": "MTo6Mw==",
    },
    Object {
      "id": "Mjo6Mw==",
    },
    Object {
      "id": "Mzo6Mw==",
    },
    Object {
      "id": "NDo6Mw==",
    },
    Object {
      "id": "NTo6Mw==",
    },
    Object {
      "id": "Njo6Mw==",
    },
    Object {
      "id": "Nzo6Mw==",
    },
    Object {
      "id": "ODo6Mw==",
    },
    Object {
      "id": "OTo6Mw==",
    },
    Object {
      "id": "MDo6NA==",
    },
    Object {
      "id": "MTo6NA==",
    },
    Object {
      "id": "Mjo6NA==",
    },
    Object {
      "id": "Mzo6NA==",
    },
    Object {
      "id": "NDo6NA==",
    },
    Object {
      "id": "NTo6NA==",
    },
  ],
  "pageInfo": Object {
    "endCursor": "NTo6NA==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
    "hasNextPage": true,
    "hasPreviousPage": true,
    "startCursor": "MTo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
  },
  "totalCount": 100,
}
`;

exports[`connectionResolver when the first [X] number are requested when there is a fresh cache and there are NO missing pages in the cache 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "MDo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "MDo6MQ==",
      },
    },
    Object {
      "cursor": "MTo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "MTo6MQ==",
      },
    },
    Object {
      "cursor": "Mjo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Mjo6MQ==",
      },
    },
    Object {
      "cursor": "Mzo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "Mzo6MQ==",
      },
    },
    Object {
      "cursor": "NDo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
      "node": Object {
        "id": "NDo6MQ==",
      },
    },
  ],
  "errors": Array [],
  "nodes": Array [
    Object {
      "id": "MDo6MQ==",
    },
    Object {
      "id": "MTo6MQ==",
    },
    Object {
      "id": "Mjo6MQ==",
    },
    Object {
      "id": "Mzo6MQ==",
    },
    Object {
      "id": "NDo6MQ==",
    },
  ],
  "pageInfo": Object {
    "endCursor": "NDo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
    "hasNextPage": true,
    "hasPreviousPage": false,
    "startCursor": "MDo6MQ==::eyJxdWVyeSI6IkhlbGxvIHdvcmxkISJ9",
  },
  "totalCount": 100,
}
`;

exports[`connectionResolver when the first [X] number are requested when there is a fresh cache and there are missing pages in the cache 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "TURvNk1RPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "MDo6MQ==",
      },
    },
    Object {
      "cursor": "TVRvNk1RPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "MTo6MQ==",
      },
    },
    Object {
      "cursor": "TWpvNk1RPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "Mjo6MQ==",
      },
    },
    Object {
      "cursor": "TXpvNk1RPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "Mzo6MQ==",
      },
    },
    Object {
      "cursor": "TkRvNk1RPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
      "node": Object {
        "id": "NDo6MQ==",
      },
    },
  ],
  "errors": Array [],
  "nodes": Array [
    Object {
      "id": "MDo6MQ==",
    },
    Object {
      "id": "MTo6MQ==",
    },
    Object {
      "id": "Mjo6MQ==",
    },
    Object {
      "id": "Mzo6MQ==",
    },
    Object {
      "id": "NDo6MQ==",
    },
  ],
  "pageInfo": Object {
    "endCursor": "TkRvNk1RPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
    "hasNextPage": true,
    "hasPreviousPage": false,
    "startCursor": "TURvNk1RPT06OnsicXVlcnkiOiJIZWxsbyB3b3JsZCEifQ==",
  },
  "totalCount": 100,
}
`;
